bin/ch-completion.bash

```diff
@@ -0,0 +1,388 @@
+        ;;
+    build-cache )
+        COMPREPLY=( $(compgen -W "--reset --gc --tree --dot" -- $cur) )
+        ;;
+    delete | list)
+        # FIXME: Is it janky to include “list” here? I'm leaning towards
+        #        “not that janky,” others may disagree.
+        if [[ "$sub_cmd" == "list" ]]; then
+            extras="$extras -l --long"
+        fi
+        # The following check seems to fix a bug where the completion
+        # function initialzes an empty storage directory.
+        if [[ -n "$(ls $strg_dir/img)" ]]; then
+            COMPREPLY=( $(compgen -W "$($CH_BIN/ch-image list -s $strg_dir) $extras" -- $cur) )
+            __ltrim_colon_completions "$cur"
+        fi
+        ;;
+    gestalt )
+        COMPREPLY=( $(compgen -W "bucache bucache-dot python-path
+                                  storage-path" -- $cur) )
+        ;;
+    import )
+        # Complete dirs and files matching the globs “*.tar.*” and “*.tgz”
+        # (a.k.a. tarballs).
+        COMPREPLY+=( $(_compgen_filepaths -X "!*.tar.* !*tgz" "$cur") )
+        if [[ ${#COMPREPLY} -gt 0 ]]; then
+            compopt -o nospace
+        fi
+        ;;
+    push )
+        if [[ "$prev" == "--image" ]]; then
+            compopt -o nospace
+            COMPREPLY=( $(compgen -d -S / -- $cur) )
+            return 0
+        # The following check seems to fix a bug where the completion
+        # function initialzes an empty storage directory.
+        elif [[ -n "$(ls "$strg_dir"/img)" ]]; then
+            COMPREPLY=( $(compgen -W "$($CH_BIN/ch-image list -s $strg_dir) --image" -- "$cur") )
+            _ltrim_colon_completions "$cur"
+        fi
+        ;;
+    undelete )
+        # FIXME: Update with “ch-image undelete --list” once #1551 drops
+        COMPREPLY=()
+        ;;
+    '' )
+        # Only autocomplete subcommands if there's no subcommand present.
+        COMPREPLY=( $(compgen -W "$_image_subcommands" -- $cur) )
+        ;;
+    esac
+
+    # If we've made it this far, the last remaining option for completion
+    # is common opts. Note that we do the “-n” check to avoid being overzealous
+    # with our suggestions.
+    if [[ -n "$cur" ]]; then
+        COMPREPLY+=( $(compgen -W "$_image_common_opts" -- $cur) )
+        return 0
+    fi
+}
+
+## ch-run ##
+
+# Completion function for ch-run
+#
+_ch_run_completion () {
+    echo "does nothing"
+}
+
+## Helper functions ##
+
+# Use this function to enable or disable completion after sourcing
+# this file. Useful if there are issues.
+# FIXME: Add other executables as their completion functions are
+#        implemented.
+ch-completion () {
+    if [[ "$1" == "disable" ]]; then
+        complete -r ch-image
+    elif [[ "$1" == "enable" ]]; then
+        complete -F _ch_image_completion ch-image
+    fi
+}
+
+# Figure out which storage directory to use (including cli-specified
+# storage).
+# FIXME: Can probably cook up a sed pattern that'll remove the need
+#        for the “if” statement.
```

Comment:
I think this might be hard because you want the replacement to differ based on which part of the pattern matched. But, you can simplify the `grep` I think:

```
$ echo 'ch-image -s foo -a' | grep -Eq -- '\s(--storage|-\w*s)'; echo $?
0
$ echo 'ch-image -as foo -a' | grep -Eq -- '\s(--storage|-\w*s)'; echo $?
0
$ echo 'ch-image --storage foo -a' | grep -Eq -- '\s(--storage|-\w*s)'; echo $?
0
$ echo 'ch-image -b foo -a' | grep -Eq -- '\s(--storage|-\w*s)'; echo $?
1
```